File: src/main/java/org/moinex/app/AppStartupRunner.java
package org.moinex.app;

import java.util.concurrent.CompletableFuture;
import org.moinex.repository.investment.TickerPurchaseRepository;
import org.moinex.repository.investment.TickerSaleRepository;
import org.moinex.service.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.stereotype.Component;

@Component
public class AppStartupRunner implements ApplicationRunner {

    private static final Logger logger = LoggerFactory.getLogger(AppStartupRunner.class);

    private final MarketService marketService;
    private final RecurringTransactionService recurringTransactionService;
    private final TickerPriceHistoryService priceHistoryService;
    private final TickerService tickerService;
    private final TickerPurchaseRepository tickerPurchaseRepository;
    private final TickerSaleRepository tickerSaleRepository;
    private final InvestmentPerformanceCalculationService investmentPerformanceCalculationService;
    private final MarketIndicatorService marketIndicatorService;
    private final BondInterestCalculationService bondInterestCalculationService;

    @Autowired
    public AppStartupRunner(
            MarketService marketService,
            RecurringTransactionService recurringTransactionService,
            TickerPriceHistoryService priceHistoryService,
            TickerService tickerService,
            TickerPurchaseRepository tickerPurchaseRepository,
            TickerSaleRepository tickerSaleRepository,
            InvestmentPerformanceCalculationService investmentPerformanceCalculationService,
            MarketIndicatorService marketIndicatorService,
            BondInterestCalculationService bondInterestCalculationService) {
        this.marketService = marketService;
        this.recurringTransactionService = recurringTransactionService;
        this.priceHistoryService = priceHistoryService;
        this.tickerService = tickerService;
        this.tickerPurchaseRepository = tickerPurchaseRepository;
        this.tickerSaleRepository = tickerSaleRepository;
        this.investmentPerformanceCalculationService = investmentPerformanceCalculationService;
        this.marketIndicatorService = marketIndicatorService;
        this.bondInterestCalculationService = bondInterestCalculationService;
    }

    @Override
    public void run(ApplicationArguments args) {
        recurringTransactionService.processRecurringTransactions();

        CompletableFuture.runAsync(
                        () -> {
                            logger.info("Updating market indicator history...");
                            marketIndicatorService.updateAllIndicators();
                            logger.info("Market indicator history updated successfully");
                        })
                .exceptionally(
                        ex -> {
                            logger.warn(
                                    "Failed to update market indicator history: {}",
                                    ex.getMessage());
                            return null;
                        })
                .thenRun(
                        () -> {
                            logger.info("Calculating bond interest...");
                            bondInterestCalculationService.calculateInterestForAllBondsIfNeeded();
                            logger.info("Bond interest calculation completed successfully");
                        })
                .exceptionally(
                        ex -> {
                            logger.warn("Failed to calculate bond interest: {}", ex.getMessage());
                            return null;
                        })
                .thenCompose(
                        ignored ->
                                marketService
                                        .updateBrazilianMarketIndicatorsFromApiAsync()
                                        .exceptionally(
                                                ex -> {
                                                    logger.error(
                                                            "Failed to update Brazilian market"
                                                                    + " indicators",
                                                            ex);
                                                    return null;
                                                })
                                        .thenCompose(
                                                ignored2 ->
                                                        marketService
                                                                .updateMarketQuotesAndCommoditiesFromApiAsync()
                                                                .exceptionally(
                                                                        ex -> {
                                                                            logger.error(
                                                                                    "Failed to"
                                                                                        + " update"
                                                                                        + " market"
                                                                                        + " quotes"
                                                                                        + " and commodities",
                                                                                    ex);
                                                                            return null;
                                                                        }))
                                        .thenCompose(
                                                ignored3 ->
                                                        priceHistoryService
                                                                .initializePriceHistory(
                                                                        tickerPurchaseRepository,
                                                                        tickerSaleRepository)
                                                                .handle(
                                                                        (res, ex) -> {
                                                                            if (ex != null) {
                                                                                logger.error(
                                                                                        "Failed to"
                                                                                            + " initialize"
                                                                                            + " price"
                                                                                            + " history",
                                                                                        ex);
                                                                                return false;
                                                                            }
                                                                            return true;
                                                                        }))
                                        .thenCompose(
                                                previousStepSuccess ->
                                                        tickerService
                                                                .updateAllNonArchivedTickersPricesAsync()
                                                                .handle(
                                                                        (res, ex) -> {
                                                                            if (ex != null) {
                                                                                logger.error(
                                                                                        "Failed to"
                                                                                            + " update"
                                                                                            + " ticker"
                                                                                            + " prices",
                                                                                        ex);
                                                                                return false;
                                                                            }
                                                                            return previousStepSuccess;
                                                                        }))
                                        .thenCompose(
                                                bothSucceeded -> {
                                                    if (Boolean.TRUE.equals(bothSucceeded)) {
                                                        return investmentPerformanceCalculationService
                                                                .recalculateAllSnapshots();
                                                    }
                                                    return CompletableFuture.completedFuture(null);
                                                })
                                        .exceptionally(
                                                ex -> {
                                                    logger.error(
                                                            "Unexpected error in startup pipeline",
                                                            ex);
                                                    return null;
                                                }));
    }
}

File: src/main/kotlin/org/moinex/app/AppStartupRunner.kt
/*
 * Filename: AppStartupRunner.kt (original filename: AppStartupRunner.java)
 * Created on: May 4, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 10/03/2026
 */

package org.moinex.app

import jakarta.annotation.PreDestroy
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import org.moinex.common.util.APIUtils
import org.moinex.common.util.FxUtils
import org.moinex.service.creditcard.RecurringCreditCardDebtService
import org.moinex.service.investment.BondInterestCalculationService
import org.moinex.service.investment.InvestmentPerformanceService
import org.moinex.service.investment.MarketIndicatorService
import org.moinex.service.investment.MarketService
import org.moinex.service.investment.TickerPriceHistoryService
import org.moinex.service.investment.TickerService
import org.moinex.service.wallet.RecurringTransactionService
import org.slf4j.LoggerFactory
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.stereotype.Component

@Component
class AppStartupRunner(
    private val marketService: MarketService,
    private val recurringTransactionService: RecurringTransactionService,
    private val recurringCreditCardDebtService: RecurringCreditCardDebtService,
    private val tickerPriceHistoryService: TickerPriceHistoryService,
    private val tickerService: TickerService,
    private val investmentPerformanceService: InvestmentPerformanceService,
    private val marketIndicatorService: MarketIndicatorService,
    private val bondInterestCalculationService: BondInterestCalculationService,
) : ApplicationRunner {
    private val logger = LoggerFactory.getLogger(AppStartupRunner::class.java)

    @Volatile
    private var isShuttingDown = false

    @PreDestroy
    fun onShutdown() {
        isShuttingDown = true
        logger.info("AppStartupRunner shutdown signal received")
    }

    override fun run(args: ApplicationArguments) {
        logger.info("AppStartupRunner started")

        FxUtils.launchOnBackground {
            logger.info("Processing recurring wallet transactions")
            recurringTransactionService.processRecurringTransactions()

            logger.info("Processing recurring credit card debts")
            recurringCreditCardDebtService.processRecurringDebts()

            logger.info("Startup pipeline beginning")
            startupPipeline()
            logger.info("Startup pipeline completed successfully")
        }
    }

    private suspend fun startupPipeline() =
        coroutineScope {
            val marketIndicatorStep =
                async {
                    runStep("Updating market indicator history") {
                        marketIndicatorService.updateAllIndicators()
                    }
                }

            val brazilianMarketIndicatorsStep =
                async {
                    runStep("Updating Brazilian market indicators") {
                        marketService.updateBrazilianMarketIndicatorsFromApi()
                    }
                }

            val quotesAndCommoditiesStep =
                async {
                    runStep("Updating market quotes and commodities") {
                        marketService.updateMarketQuotesAndCommoditiesFromApi()
                    }
                }

            marketIndicatorStep.await()
            brazilianMarketIndicatorsStep.await()
            quotesAndCommoditiesStep.await()

            val bondInterestCalculationStep =
                async {
                    runStep("Calculating bond interest") {
                        bondInterestCalculationService.calculateInterestForAllBondsIfNeeded()
                    }
                }

            val tickerPriceHistoryStep =
                async {
                    runStep("Initializing price history") {
                        tickerPriceHistoryService.initializePriceHistory()
                    }
                }

            val nonArchivedTickerPricesStep =
                async {
                    runStep("Updating ticker prices") {
                        tickerService.updateAllNonArchivedTickersPrices()
                    }
                }

            bondInterestCalculationStep.await()
            tickerPriceHistoryStep.await()
            nonArchivedTickerPricesStep.await()

            runStep("Recalculating investment performance snapshots") {
                investmentPerformanceService.recalculateAllSnapshots()
            }
        }

    private suspend fun <T> runStep(
        name: String,
        block: suspend () -> T,
    ): T? {
        if (isShuttingDown) {
            logger.info("{} skipped: Application is shutting down", name)
            return null
        }

        return runCatching {
            logger.info("Starting {}", name)
            val result = block()

            if (isShuttingDown || APIUtils.isShuttingDown()) {
                logger.info("{} completed but application is shutting down", name)
                return null
            }

            logger.info("{} completed successfully", name)
            result
        }.getOrElse { e ->
            if (isShuttingDown || APIUtils.isShuttingDown()) {
                logger.info("{} cancelled: Application is shutting down", name)
            } else {
                logger.error("{} failed", name, e)
            }
            null
        }
    }
}
